lib/cartopy/io/shapereader.py

```diff
@@ -234,6 +253,83 @@ def records(self):
                          fields)
 
 
+class FionaReader(object):
+    """
+    Provides an interface for accessing the contents of a shapefile
+    with the fiona library, which has a much faster reader than pyshp.
+
+    The primary methods used on a Reader instance are
+    :meth:`~Reader.records` and :meth:`~Reader.geometries`.
+
+    """
+    def __init__(self, filename, bbox=None):
+        self._data = []
+
+        with fiona.open(filename) as f:
+            if bbox is not None:
+                assert len(bbox) == 4
+                features = f.filter(bbox=bbox)
+            else:
+                features = f
+
+            # Handle feature collections
+            if hasattr(features, "__geo_interface__"):
+                fs = features.__geo_interface__
+            else:
+                fs = features
+
+            if isinstance(fs, dict) and fs.get('type') == 'FeatureCollection':
+                features_lst = fs['features']
+            else:
+                features_lst = features
+
+            for feature in features_lst:
+                if hasattr(f, "__geo_interface__"):
+                    feature = feature.__geo_interface__
+                else:
+                    feature = feature
+
+                d = {'geometry': sgeom.shape(feature['geometry'])
+                     if feature['geometry'] else None}
+                d.update(feature['properties'])
+                self._data.append(d)
+
+    def __len__(self):
+        return len(self._data)
+
+    def geometries(self):
+        """
+        Returns an iterator of shapely geometries from the shapefile.
+
+        This interface is useful for accessing the geometries of the
+        shapefile where knowledge of the associated metadata is desired.
+        In the case where further metadata is needed use the
+        :meth:`~Reader.records`
+        interface instead, extracting the geometry from the record with the
+        :meth:`~Record.geometry` method.
+
+        """
+        for i in range(len(self)):
```

Comment:
Wouldn't this be slightly cleaner and faster as:
```python
for item in self._data:
    yield item['geometry']
```

Reply:
@dopplershift, Thanks for finding that! I've updated another spot below that as well. Too much copying from the BasicReader that was already implemented. Might want to refactor the design of the BasicReader to not require indexing in the future.

I was also struggling with the rebasing, but think I have it all updated now along with updating the "What's New" section.